src/klein/_headers.py

```diff
@@ -0,0 +1,356 @@
+        convert=normalizeRawHeadersMutable
+    )  # type: MutableRawHeaders
+
+
+    @property
+    def rawHeaders(self):
+        # type: () -> RawHeaders
+        return tuple(self._rawHeaders)
+
+
+    def getValues(self, name):
+        # type: (AnyStr) -> Iterable[AnyStr]
+        return getFromRawHeaders(self._rawHeaders, name)
+
+
+    def remove(self, name):
+        # type: (String) -> None
+        if type(name) is bytes:
+            rawName = name
+        elif type(name) is Text:
+            rawName = headerNameAsBytes(name)
+        else:
+            raise TypeError("name must be text or bytes")
+
+        self._rawHeaders[:] = [p for p in self._rawHeaders if p[0] != rawName]
+
+
+    def addValue(self, name, value):
+        # type: (AnyStr, AnyStr) -> None
+        if type(name) is bytes:
+            if type(value) is not bytes:
+                raise TypeError("value must be bytes to match name")
+
+            rawName  = name   # type: bytes
+            rawValue = value  # type: bytes
+
+        elif type(name) is Text:
+            if type(value) is not Text:
+                raise TypeError("value must be text to match name")
+
+            rawName  = headerNameAsBytes(name)
+            rawValue = headerValueAsBytes(value)
+
+        else:
+            raise TypeError("name must be text or bytes")
+
+        self._rawHeaders.append((rawName, rawValue))
+
+
+
+# Support for L{Headers}
+
+@implementer(IHTTPHeaders)
+@attrs(frozen=True)
+class HTTPHeadersFromHeaders(object):
+    """
+    HTTP entity headers.
+
+    This is an L{IHTTPHeaders} implementation that wraps a L{Headers}
+    object.
+
+    This is used by Klein to expose objects from L{twisted.web} to clients
+    while presenting the Klein interface.
+    """
+
+    @attrs(frozen=False)
+    class _State(object):
+        """
+        Internal mutable state for L{HTTPRequestFromIRequest}.
+        """
+
+    _headers = attrib(validator=instance_of(Headers))     # type: Headers
+    _state = attrib(default=Factory(_State), init=False)  # type: _State
+
+
+    @property
+    def rawHeaders(self):
+        # type: () -> RawHeaders
+        def pairs():
+            # type: () -> Iterable[Tuple[bytes, bytes]]
+            for name, values in self._headers.getAllRawHeaders():
+                name = normalizeHeaderName(name)
+                for value in values:
+                    yield (name, value)
+
+        return tuple(pairs())
+
+
+    def getValues(self, name):
+        # type: (AnyStr) -> Iterable[AnyStr]
+        if type(name) is bytes:
+            values = self._headers.getRawHeaders(name, default=())
+        elif type(name) is Text:
+            values = (
+                headerValueAsText(value)
+                for value in self._headers.getRawHeaders(
+                    headerNameAsBytes(name), default=()
+                )
+            )
+        else:
+            raise TypeError("name must be text or bytes")
+
+        return tuple(values)
```

Comment:
I might be misunderstanding the pattern of public/private module naming; but it seems like much of the headers code should be private API.